### ISSUE TITLE
Seems the sampled arrays or covariance matrix is somehow permuted

### ISSUE THREAD

**Issue by sinaasadiyan** [brunzema/truncated-mvn-sampler#5]
I came across testing this implementation. First I thought the results are wrong but when I tried multiple times for 20000 sampled points I figured out that the variance of generated arrays is not compatible with the covariance matrix. When I printed the variance I noticed the position of the elements are not consistent. So the mean of array is Ok but the variance of the array is mistaken by another array. So the result is wrong.

Am I right?

![Screenshot (562)](https://github.com/brunzema/truncated-mvn-sampler/assets/7326608/5481289a-83c2-4799-a55e-e65e564cdb12)

**Comment by brunzema** [comment]
Hi, can you give the full examples including the bounds that you specified?

The following code gives me the expected behavior:
```python
np.random.seed(0)

d = 5  # dimensions

# random mu and cov
mu = np.random.rand(d)
cov = 0.5 - np.random.rand(d ** 2).reshape((d, d))
cov = np.triu(cov)
cov += cov.T - np.diag(cov.diagonal())
cov = np.dot(cov, cov)

# constraints
lb = -np.ones_like(mu) * np.inf
ub = np.ones_like(mu) * np.inf

# create truncated normal and sample from it
n_samples = 100000
tmvn = TruncatedMVN(mu, cov, lb, ub)
samples = tmvn.sample(n_samples)

print(f"True mean: {mu}")
print(f"Sample mean: {np.mean(samples, axis=1)}")

print(f"True cov diag: {np.diag(cov)}")
print(f"Sample var: {np.var(samples, axis=1)}")
```

Output:
```
True mean: [0.5488135  0.71518937 0.60276338 0.54488318 0.4236548 ]
Sample mean: [0.54976451 0.71808884 0.60040936 0.54846622 0.4248551 ]
True cov diag: [0.40723557 0.3745029  0.48303228 0.69789786 0.48750193]
Sample var: [0.40503953 0.37333319 0.48350073 0.69871607 0.48725899]
```

Also not that once you enforce the bounds, the sample mean and variance can significantly differ from the mean and variance of the truncated normal (depending on your bounds) since you are cutting off the probability mass in the tails.

Hope this is helpful :)

**Comment by sinaasadiyan** [comment]
[MTND.zip](https://github.com/brunzema/truncated-mvn-sampler/files/14957619/MTND.zip)

You are right. I checked with not truncated constraints (+-inf) and that is working. 
I had tested multiple times and as mean values were equal and permuted variances (as I thought) were correct with 2 digits of floating, I thought that is a bug 🤦‍♂️. 

Maybe my bounds are not zero centered or generated random variance/covariance values are not ok.
I attached my code. What is the best algorithm for this task?

Thanks for this implementation. I closed this issue.
